24.0.0.5-beta staging (#3742)
OpenLiberty/blogs: posts/2024-05-07-24.0.0.5-beta.adoc

---
layout: post
title: "Jakarta EE 11 previews and InstantOn support for IBM MQ in 24.0.0.5-beta"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/dmuelle
author_github: https://github.com/dmuelle
seo-title: Jakarta EE 11 previews and InstantOn support for IBM MQ in 24.0.0.5-beta - OpenLiberty.io

> seo-description: The 24.0.0.5-beta release introduces previews for several Jakarta EE 11 features, including Jakarta Concurrency 3.1, Jakarta Data 1.0, Jakarta Expression Language 6.0, Jakarta Pages 4.0, and Jakarta Servlet 6.1. This release also includes support for using InstantOn with IBM MQ messaging.

Reviewer: ```suggestion
seo-description: The 24.0.0.5-beta release introduces previews for several Jakarta EE 11 features, including Jakarta Contexts and Dependency Injection 4.1, Jakarta Concurrency 3.1, Jakarta Data 1.0, Jakarta Expression Language 6.0, Jakarta Pages 4.0, and Jakarta Servlet 6.1. This release also includes support for using InstantOn with IBM MQ messaging.

```

blog_description: The 24.0.0.5-beta release introduces previews for several Jakarta EE 11 features, including Jakarta Concurrency 3.1, Jakarta Data 1.0, Jakarta Expression Language 6.0, Jakarta Pages 4.0, and Jakarta Servlet 6.1. This release also includes support for using InstantOn with IBM MQ messaging.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Jakarta EE 11 previews and InstantOn support for IBM MQ in 24.0.0.5-beta
David Mueller <https://github.com/dmuelle>
:imagesdir: /
:url-prefix:
:url-about: /

> The 24.0.0.5-beta release introduces previews for several Jakarta EE 11 features, including Jakarta Concurrency 3.1, Jakarta Data 1.0, Jakarta Expression Language 6.0, Jakarta Pages 4.0, and Jakarta Servlet 6.1. This release also includes support for using InstantOn with IBM MQ messaging.

Reviewer 2: CDI 4.1 appears in a subsequent listing, but is missing from this paragraph.
```suggestion
The 24.0.0.5-beta release introduces previews for several Jakarta EE 11 features, including Jakarta Contexts and Dependency Injection 4.1, Jakarta Concurrency 3.1, Jakarta Data 1.0, Jakarta Expression Language 6.0, Jakarta Pages 4.0, and Jakarta Servlet 6.1. This release also includes support for using InstantOn with IBM MQ messaging.

```

* link:https://jakarta.ee/specifications/cdi/4.1/apidocs/[CDI API]
* link:https://jakarta.ee/specifications/interceptors/2.2/apidocs/[Interceptors API]
* link:https://jakarta.ee/specifications/annotations/3.0/apidocs/jakarta.annotation/module-summary.html[Jakarta Annotations API]
* link:https://jakarta.ee/specifications/cdi/4.1/jakarta-cdi-spec-4.1[CDI Specification]
* link:https://jakarta.ee/specifications/interceptors/2.2/jakarta-interceptors-spec-2.2[Interceptors Specification]
* link:https://jakarta.ee/specifications/annotations/3.0/annotations-spec-3.0[Jakarta Annotations Specification]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/28075
// Contact/Reviewer: mswatosh
// // // // // // // //
[#concurrent]
== Jakarta Concurrency 3.1

Jakarta Concurrency allows applications to use concurrency while maintaining the benefits of running on a Jakarta EE runtime. Jakarta Concurrency 3.1 is the new version for the upcoming Jakarta EE 11, and it adds support for new features of the latest Java SE releases as well as some usability improvements.

> The new features in Jakarta Concurrency 3.1 are improved support for the Java Flow (Reactive Streams) APIs, and the use of virtual threads in `ManagedExecutors` and `ManagedThreadFactories`. Support is also included for using `@Inject` in place of `@Resource`, and the option to run asynchronous methods on a schedule by using the new `@Schedule` annotation.

Reviewer 2: Corrections to API class names and grammar fix to commas.
```suggestion
The new features in Jakarta Concurrency 3.1 are improved support for the Java Flow (Reactive Streams) APIs and the use of virtual threads in managed executors and managed thread factories. Support is also included for using `@Inject` in place of `@Resource` and the option to run asynchronous methods on a schedule by using the new `@Schedule` annotation.

```

> To support the Flow APIs, `ContextService` has two new methods, `contextualSubscriber`, which provides context in a `Flow.Subscriber`, and `Flow.Processor`, which provides context in a `contextualProcessor`. These new methods allow for context in a flow when using a `Flow.Publisher` which does not allow configuring a `ManagedExecutor`.

Reviewer 2: The second method name/class name combination is backwards.
```suggestion
To support the Flow APIs, `ContextService` has two new methods, `contextualSubscriber`, which provides context in a `Flow.Subscriber`, and `contextualProcessor`, which provides context in a `Flow.Processor`. These new methods allow for context in a flow when using a `Flow.Publisher` that does not allow configuring a `ManagedExecutor`.

```

Reviewer: sorry, missed this. Fixed now.

[source,java]
----
publisher.subscribe(contextService.contextualSubscriber(subscriber));
----

> Virtual threads are available in Jakarta Concurrency 3.1 by specifying ‘virtual = true’ on a `ManagedExecutorDefinition`, `ManagedScheduledExecutorDefinition`, or `ManagedThreadFactory` instance. Virtual threads are a new feature of Java 21, so when running on Java 17 if ‘virtual = true’ is specified, platform threads are provided instead of virtual threads.

Reviewer 2: I think it reads better like this. Feel free to ignore the comment if you don't agree.
```suggestion
Virtual threads are available in Jakarta Concurrency 3.1 by specifying ‘virtual = true’ on a `ManagedExecutorDefinition`, `ManagedScheduledExecutorDefinition`, or `ManagedThreadFactory` instance. Virtual threads are a new feature of Java 21, so when running on Java 17 and ‘virtual = true’ is specified, platform threads are provided instead of virtual threads.

```

[source,java]
----
@ManagedExecutorDefinition(name = "java:module/concurrent/virtual-executor",
                           virtual = true)
----

The new `@Schedule` annotation allows asynchronous methods to run on a schedule. In this example, after the method is called, it runs asynchronously on a schedule set by the provided cron string

[source,java]
----
@Asynchronous(runAt = { @Schedule(cron = "*/3 * * * * *")})
void scheduledMethod() {

> ----

Reviewer 2: I think this example must have been copied from a test case. It needs to be explained what it achieves. Running every 3 seconds is not a likely customer scenario and we should choose an example that is more meaningful.
```suggestion
The new `@Schedule` annotation allows asynchronous methods to run on a schedule. In this example, after the method is called, it runs asynchronously at 10 AM on every day of every month where the day of the week is a Monday, Wednesday, or Friday, according to the provided cron string.

[source,java]
----
@Asynchronous(runAt = @Schedule(cron = "0 10 * * MON,WED,FRI", zone = "America/Chicago"))
void runAt10MWF() {
    ...
    if (neverRunAgain)
        Asynchronous.Result.complete(null);
}
----

```

Reviewer 3: I used what I had in my demo, which needed to work at any time, I agree this is better for a blog post.

     <feature>Data-1.0</feature>
</featureManager>`
----

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/27964
// Contact/Reviewer: pnicolucci
// // // // // // // //
[#el]
== Jakarta Expression Language 6.0

The `expressionLanguage-6.0` feature is an implementation of the Expression Language 6.0 Specification for Jakarta EE 11. The Expression Language 6.0 specification includes a number of new features and specification clarifications.

Support for `java.util.Optional` (not enabled by default) and `java.lang.Record` is added. Another new feature is the addition of the length property for Arrays. For more information and the change history of the specification between Expression Language 5.0 and Expression Language 6.0, see link:https://jakarta.ee/specifications/expression-language/6.0/jakarta-expression-language-spec-6.0#changes-between-6-0-and-5-0[Changes between 6.0 and 5.0].

To use this feature, add the following code to your `server.xml` file:

[source,xml]
----
<featureManager>
     <feature>expressionLanguage-6.0</feature>

> </featureManager>`

Reviewer 4: Remove the  \` after `</featureManager>`

The `jakarta.servlet.jsp.ErrorData` class was updated to add support for the new `jakarta.servlet.error.method` and `jakarta.servlet.error.query_string` attributes.  The following deprecated classes, methods, and actions were removed:

* `jakarta.servlet.jsp.JspException.getRootCause()`
* Classes in `jakarta.servlet.jsp.el`.
* `isThreadSafe` page directive
* `jsp:plugin` action and related actions
* `jakarta.servet.jsp.tagext.BodyTag.EVAL_BODY_TAG` constant
* Any methods that implemented `jakarta.el.ELResolver.getFeatureDescriptors()` were removed as the `getFeatureDescriptors()` method was removed in Expression Language 6.0.

For more information and the change history of the specification between Pages 3.1 and Pages 4.0, see link:https://github.com/jakartaee/pages/blob/master/spec/src/main/asciidoc/ServerPages.adoc#c1-changes-between-jsp-40-and-jsp-31[Changes between 3.1 and 4.0].

To use this feature, add the following code to your `server.xml` file:

[source,xml]
----
<featureManager>
     <feature>pages-4.0</feature>

> </featureManager>`

Reviewer 4: Remove the  \` after `</featureManager>`

----

For more information, see the https://jakarta.ee/specifications/pages/4.0/[Jakarta Pages 4.0 Specification].

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/27962
// Contact/Reviewer: pmd1nh,pnicolucci
// // // // // // // //
[#servlet]
== Jakarta Servlet 6.1

The Open Liberty `servlet-6.1` feature is an implementation of the Servlet 6.1 specification for Jakarta EE 11 . It includes a number of new features, specification clarifications, and deprecates some existing servlet features.`

> 

Reviewer 5: Remove the extra \` at the end.

Reviewer: fixed

Before Servlet 6.1, there was no way for an application to control the response data when doing a send redirect, as well as the response status code, which always set to 302. An application could not easily retrieve the initial request's query string or request HTTP method during an error handling dispatch. It also could not set the character encoding for a request or a response using the `java.nio.charset.Charset`; the only available option was using a String.  During the read or write of a servlet data, the `jakarta.servlet.ServletInputStream` or `jakarta.servlet.ServletOutputStream` only supports a byte array.

Servlet 6.1 provides servlet APIs that allow the send redirect to include an optional response data or set a compliant status code, instead of the default 302. Additional request attributes are available during the error handling process to easily retrieve the initial request’s query string or method. Furthermore, international applications can now set the character encoding using a Charset instead of a String.  The `ServletInputStream` or `ServletOutputStream` can use `java.nio.ByteBuffer` to read or write the data.

> Several clarifications are added to the behavior of the existing servlet APIs. For example, the `getParameter` family from the `jakar.servlet.ServletRequest` is now throwing the runtime j`ava.lang.IllegalStateException` when an error occurs during the parsing of the request’s parameters.  While processing an error-handling dispatch, the HTTP GET method is always used instead of the original request’s HTTP method.`

Reviewer 5: move the j inside \` ava.lang.IllegalStateException

> Several clarifications are added to the behavior of the existing servlet APIs. For example, the `getParameter` family from the `jakar.servlet.ServletRequest` is now throwing the runtime j`ava.lang.IllegalStateException` when an error occurs during the parsing of the request’s parameters.  While processing an error-handling dispatch, the HTTP GET method is always used instead of the original request’s HTTP method.`

Reviewer 5: jakar.servlet.ServletRequest = jakarta.servlet.ServletRequest

> 

Reviewer 5: extra \` after `the HTTP method.`

Reviewer: fixed

---
layout: post
title: "Jakarta EE 11 previews and InstantOn support for IBM MQ in 24.0.0.5-beta"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/dmuelle
author_github: https://github.com/dmuelle
seo-title: Jakarta EE 11 previews and InstantOn support for IBM MQ in 24.0.0.5-beta - OpenLiberty.io
seo-description: The 24.0.0.5-beta release introduces previews for several Jakarta EE 11 features, including Jakarta Contexts and Dependency Injection 4.1, Jakarta Concurrency 3.1, Jakarta Data 1.0, Jakarta Expression Language 6.0, Jakarta Pages 4.0, and Jakarta Servlet 6.1. This release also includes support for using InstantOn with IBM MQ messaging.

> blog_description: The 24.0.0.5-beta release introduces previews for several Jakarta EE 11 features, including Jakarta Concurrency 3.1, Jakarta Data 1.0, Jakarta Expression Language 6.0, Jakarta Pages 4.0, and Jakarta Servlet 6.1. This release also includes support for using InstantOn with IBM MQ messaging.

Reviewer: ```suggestion
blog_description: The 24.0.0.5-beta release introduces previews for several Jakarta EE 11 features, including Jakarta Contexts and Dependency Injection 4.1, Jakarta Concurrency 3.1, Jakarta Data 1.0, Jakarta Expression Language 6.0, Jakarta Pages 4.0, and Jakarta Servlet 6.1. This release also includes support for using InstantOn with IBM MQ messaging.

```

Here's a server configuration snippet to deploy applications that use the JakartaEE-10 Messaging API and require the link:https://repo1.maven.org/maven2/com/ibm/mq/wmq.jakarta.jmsra/9.3.5.0/wmq.jakarta.jmsra-9.3.5.0.rar[IBM® MQ resource adapter] to access external messaging resources. This configuration automatically includes the `connectors-2.1` feature, which supports the resource adapter configuration element.

[source,xml]
----
<featureManager>
   <feature>messaging-3.1</feature>
   <feature>mdb-4.0</feature>
   <feature>servlet-6.0</feature>
<featureManager/>

<resourceAdapter id="mqJms" location="${server.config.dir}/wmq.jakarta.jmsra-9.3.5.0.rar"/>
----

You can use the link:https://developer.ibm.com/tutorials/mq-connect-app-queue-manager-containers/[MQ in Container image] to stand up an MQ server that provides the following resources: queue manager `QM1`, queue `DEV.QUEUE.1`, channel `DEV.APP.SVRCONN`, and listener `SYSTEM.LISTENER.TCP.1` on port `1414`.  These resources can support a simple point-to-point messaging scenario and are named within the messaging feature configuration elements shown in the following example.

> InstantOn can dynamically update the Messaging and Connectors feature configurations to access resources that are provided by any environment in which a server is restored from checkpoint. The following messaging configuration snippet uses the hostname and MQ listener port values (variables) defined in the restore environment. When this server restores, the connection factory and endpoint activation configurations access MQ resources by using the updated hostname and port values.

Reviewer 6: I need to disambiguate this paragraph.  Please replace it with the following, which is more precise and enables me to "unmuddle" the usage of the HOSTNAME environment variable in the config snippets.  

InstantOn can dynamically update the Messaging and Connectors feature configuration elements to enable connections to external resources made available by any environment in which a server is restored from checkpoint.  In the snippet below the Messaging configuration elements use Liberty variables named `MQ_HOSTNAME` and MQ_PORT to specify the hostname and IP port of the system on which the MQ queue manager is running.  Define these variables in the `server.env` file of a checkpointed server, and when the server is restored, the queue connection factory and message endpoint activation configurations will update to the environment-specific hostname and port values for the MQ queue manager.

Reviewer: updated- with minor edits per IBM style. Lmk what you think

[source,xml]
----
   <jmsQueue id="jms/queue1" jndiName="jms/queue1">
      <properties.mqJms  baseQueueName="DEV.QUEUE.1"  baseQueueManagerName="QM1"/>
   </jmsQueue>

   <variable name="MQ_LISTENER_PORT" value="1414"/>

> 

Reviewer 6: At line 456: Change `MQ_LISTENER_PORT` to `MQ_PORT`
Add code at line 457:  `<variable name="MQ_HOSTNAME" value="localhost"/>`
Add blank newline

Reviewer: fixed

>    <jmsQueueConnectionFactory jndiName="jms/qcf1" connectionManagerRef="ConMgr7">

Reviewer 6: Please change all occurrences of `HOSTNAME` to `MQ_HOSTNAME`
Please change all occurrences of `MQ_LISTENER_PORT` to `MQ_PORT`.

Reviewer 6: @dmuelle I need the following changes applied to the entire config snippet below.

Please change all occurrences of HOSTNAME to MQ_HOSTNAME
Please change all occurrences of MQ_LISTENER_PORT to MQ_PORT.

      <properties.mqJms  hostName="${HOSTNAME}"  port="${MQ_LISTENER_PORT}"
            channel="DEV.APP.SVRCONN"  queueManager="QM1"/>
   </jmsQueueConnectionFactory>

   <jmsConnectionFactory jndiName="jms/cf1" connectionManagerRef="ConMgr1">
      <properties.mqJms  hostName="${HOSTNAME}"  port="${MQ_LISTENER_PORT}"
            channel="DEV.APP.SVRCONN"  queueManager="QM1"/>
   </jmsConnectionFactory>
    <connectionManager id="ConMgr1" maxPoolSize="10"/>

   <jmsActivationSpec id="myapp/mymdb/FVTMessageDrivenBean">
      <properties.mqJms  destinationRef="jms/queue1"  destinationType="jakarta.jms.Queue"
            transportType="CLIENT"  hostName="${HOSTNAME}"  port="${MQ_LISTENER_PORT}"
            channel="DEV.APP.SVRCONN"  queueManager="QM1"/>
   </jmsActivationSpec>
----

> Enjoy the time savings and stay tuned for upcoming announcements regarding InstantOn support for the Jakarta embedded messaging features.

Reviewer 6: Change "Jakarta embedded messaging" features to "Jakarta features".  GA for embedded messaging features is delayed for now.

Reviewer 6: @dmuelle Not critical to make the above change; just wondering whether you saw the request.  -Regards

[source,java]
----
@ManagedExecutorDefinition(name = "java:module/concurrent/virtual-executor",
                           virtual = true)
----

The new `@Schedule` annotation allows asynchronous methods to run on a schedule. In this example, after the method is called, it runs asynchronously at 10 AM on every day of every month where the day of the week is a Monday, Wednesday, or Friday, according to the provided cron string.

[source,java]
----
@Asynchronous(runAt = @Schedule(cron = "0 10 * * MON,WED,FRI", zone = "America/Chicago"))
void runAt10MWF() {
    ...
    if (neverRunAgain)
        Asynchronous.Result.complete(null);
}
----

To try out Jakarta Concurrency 3.1, check out the link:https://github.com/OpenLiberty/sample-concurrency[sample application].

To use this feature, add the following code to your `server.xml` file:

[source,xml]
----
<featureManager>

>      <feature>concurrent-3.0</feature>

Reviewer 2: ```suggestion
     <feature>concurrent-3.1</feature>

```

public class MyServlet extends HttpServlet {
    @Inject
    Products products;

    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
            throws ServletException, IOException {
        // Insert:
        Product prod = ...
        prod = products.add(prod);

        // Find the price of one product:
        price = products.getPrice(productId).orElseThrow();

        // Find all, sorted:
        List<Product> all = products.findByNameIgnoreCaseContains(searchFor, Order.by(
                                     _Product.price.desc(),
                                     _Product.name.asc(),
                                     _Product.id.asc()));

        // Find the first 20 most expensive products on sale:
        Page<Product> page1 = products.find(onSale, Order.by(_Product.price.desc(),
                                                             _Product.name.asc(),
                                                             _Product.id.asc())
                                                         .pageSize(20));
        ...
    }
}
----

To use this feature, add the following code to your `server.xml` file:

[source,xml]
----
<featureManager>

>      <feature>Data-1.0</feature>

Reviewer 2: ```suggestion
     <feature>data-1.0</feature>

```

[source,xml]
----
   <jmsQueue id="jms/queue1" jndiName="jms/queue1">
      <properties.mqJms  baseQueueName="DEV.QUEUE.1"  baseQueueManagerName="QM1"/>
   </jmsQueue>

   <variable name="MQ_PORT" value="1414"/>
   <variable name="MQ_HOSTNAME" value="localhost"/>

   <jmsQueueConnectionFactory jndiName="jms/qcf1" connectionManagerRef="ConMgr7">
      <properties.mqJms  hostName="${MQ_HOSTNAME}"  port="${MQ_PORT}"
            channel="DEV.APP.SVRCONN"  queueManager="QM1"/>
   </jmsQueueConnectionFactory>

   <jmsConnectionFactory jndiName="jms/cf1" connectionManagerRef="ConMgr1">
      <properties.mqJms  hostName="${MQ_HOSTNAME}"  port="${MQ_PORT}"
            channel="DEV.APP.SVRCONN"  queueManager="QM1"/>
   </jmsConnectionFactory>
    <connectionManager id="ConMgr1" maxPoolSize="10"/>

   <jmsActivationSpec id="myapp/mymdb/FVTMessageDrivenBean">
      <properties.mqJms  destinationRef="jms/queue1"  destinationType="jakarta.jms.Queue"
            transportType="CLIENT"  hostName="${MQ_HOSTNAME}"  port="${MQ_PORT}"
            channel="DEV.APP.SVRCONN"  queueManager="QM1"/>
   </jmsActivationSpec>
----

> Enjoy the time savings and stay tuned for upcoming announcements regarding InstantOn support for the Jakarta embedded messaging features.

Reviewer: ```suggestion
Enjoy the time savings and stay tuned for upcoming announcements regarding InstantOn support for Jakarta features.

```
